packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -1145,6 +1148,7 @@ Future<void> injectPlugins(
     }
   }
   if (webPlatform) {
+    assert(project.web.buildDir != null, 'project.web.buildDir must be set before calling injectPlugins');
```

Comment:
If web plugin injection is now fundamentally different than on other platforms (because it's per-build), we should probably split it out into a separate method; it seems like we're probably never going to actually want to call this for web at the same time as other platforms, making this a foot-gun. (Thus the need for the commented out line with explanation in the other file.) So, `injectPersistentPluginFiles` and `injectBuildTimePluginFiles`?

Or alternately, a required `injectionTime` enum argument to this single method with `setup` and `build`, and web can internally no-op when injectionTime is not `build`.

Reply:
Yes, this is a good call. The problem is that injectPlugins is called from another method that is called "ensureReadyForPlatformSpecificTooling" which seems interesting even for web (it's delegating to a sub-method of a class). The problem is that it then calls to injectPlugins, potentially for many platforms at the same time. I might remove *web* from injectPlugins at all, and just have a custom method.

Reply:
I've removed the `webPlatform` option from `injectPlugins`, and created a separate `injectBuildTimePluginFiles` method (almost identical to `injectPlugins` in signature) that lets the user inject plugin files as needed (where needed), for now only for webPlatform.